hs-bindgen/src-internal/HsBindgen/C/Fold/Type.hs

```diff
@@ -169,205 +169,111 @@ processTypeDecl' ctxt extBindings unit declCursor ty = case fromSimpleEnum $ cxt
-                then do
-                    -- Anonymous top-level declaration: nothing to do but warn, as there
-                    -- shouldn't be one in "good" code.
-                    return TypeVoid
-
-                else do
+                mFlavour <- classifyTypeDecl ctxt extBindings (ty, decl, ki, sloc)
+
+                case mFlavour of
+                  Left (AnonTopLevel replacement) ->
+                    return replacement
+                  Right (declPath, flavour) -> do
                     addTypeDeclProcessing ty $ TypeUnion declPath
-                    sloc <- liftIO $
-                        HighLevel.clang_getExpansionLocation
-                            =<< clang_getCursorLocation decl
-
-                    -- TODO: ExtBindings?
-                    liftIO (HighLevel.classifyDeclaration decl) >>= \case
-                            DeclarationOpaque ->
-                                -- opaque struct and opaque union look the same.
-                                addDecl ty $ DeclOpaqueStruct OpaqueStruct {
-                                      opaqueStructTag       = CName name'
-                                    , opaqueStructAliases   = []
-                                    , opaqueStructSourceLoc = sloc
-                                    }
-
-                            DeclarationForward _defn -> do
-                                liftIO $ panicIO "should not happen"
-
-                            DeclarationRegular -> do
-                                -- the below is TODO:
-                                sizeof    <- liftIO (clang_Type_getSizeOf  ty)
-                                alignment <- liftIO (clang_Type_getAlignOf ty)
-
-                                fields <- HighLevel.clang_visitChildren decl $ \cursor -> do
-                                    let mkCtxt :: CName -> DeclPathCtxt
-                                        mkCtxt fieldName
-                                          | anon      = DeclPathCtxtField Nothing              fieldName ctxt
-                                          | otherwise = DeclPathCtxtField (Just $ CName name') fieldName ctxt
-                                    mfield <- mkUnionField extBindings unit mkCtxt cursor
-                                    return $ Continue mfield
-
-                                addDecl ty $ DeclUnion Union
-                                    { unionDeclPath  = declPath
-                                    , unionAliases   = []
-                                    , unionSizeof    = fromIntegral sizeof
-                                    , unionAlignment = fromIntegral alignment
-                                    , unionFields    = fields
-                                    , unionSourceLoc = sloc
-                                    }
+                    case flavour of
+                      TypeDeclExternal _extId ->
+                        panicIO "external bindings for unions not implemented #537"
```

Comment:
Should this be `throwIO_TODO`? Or is `TypeDeclExternal` case here unreachable? Why we cannot do `addAlias ty $ TypeExtBinding extId` here?

Reply:
Ah, yes, it should be `throwIO_TODO`.